Framework/LiveData/src/Kafka/IKafkaStreamDecoder.cpp

```diff
@@ -298,18 +321,23 @@ IKafkaStreamDecoder::getStopOffsets(
  * then we wait for it to finish
  */
 void IKafkaStreamDecoder::waitForDataExtraction() {
-  {
-    std::unique_lock<std::mutex> readyLock(m_waitMutex);
+  std::unique_lock<std::mutex> readyLock(m_waitMutex);
+  if (m_extractWaiting) { // Might have extracted whilst we were grabbing
+                          // mutex
     m_cv.wait(readyLock, [&] { return !m_extractWaiting; });
   }
 }
 
 void IKafkaStreamDecoder::waitForRunEndObservation() {
-  m_extractWaiting = true;
+  {
+    std::lock_guard lck(m_waitMutex);
+    m_extractWaiting = true;
+  }
   // Mark extractedEndRunData true before waiting on the extraction to ensure
   // an immediate request for run status after extracting the data will return
   // the correct value - avoids race condition in MonitorLiveData and tests
   m_extractedEndRunData = true;
+  m_cbIterationEnd();
```

Comment:
Since we now run lockstep the main test thread will wait for some sort of signal from Kafka to carry on to the next step.

`waitForRunObservation` will also wait for the test thread to call `extractData` in a wait too, so we deadlock.

Originally I tried getting away with simply running both the test thread freely (not in lock step), but this raced and caused the failure we saw on Ubuntu.
With the mechanism the test thread is now aware that Kafka is about to enter a wait block so can resume execution. The header has a comment which states the callbacks are for unit tests, hence I felt safe doing this change.

Reply:
Ah thanks. I had missed the comment about the callbacks being for testing - this is fine. I was worried we wold end up having somewhere else receive extra callbacks it might not be expecting.